// @graphql-eslint/eslint-plugin@4.4.0/cjs/rules/require-deprecation-date/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }var _indexjs = require('../../estree-converter/index.js');
var _utilsjs = require('../../utils.js');
const DATE_REGEX = /^\d{2}\/\d{2}\/\d{4}$/;
const MESSAGE_REQUIRE_DATE = "MESSAGE_REQUIRE_DATE";
const MESSAGE_INVALID_FORMAT = "MESSAGE_INVALID_FORMAT";
const MESSAGE_INVALID_DATE = "MESSAGE_INVALID_DATE";
const MESSAGE_CAN_BE_REMOVED = "MESSAGE_CAN_BE_REMOVED";
const schema = {
  type: "array",
  maxItems: 1,
  items: {
    type: "object",
    additionalProperties: false,
    properties: {
      argumentName: {
        type: "string"
      }
    }
  }
};
const rule = {
  meta: {
    type: "suggestion",
    hasSuggestions: true,
    docs: {
      category: "Schema",
      description: "Require deletion date on `@deprecated` directive. Suggest removing deprecated things after deprecated date.",
      url: "https://the-guild.dev/graphql/eslint/rules/require-deprecation-date",
      examples: [
        {
          title: "Incorrect",
          code: (
            /* GraphQL */
            `
            type User {
              firstname: String @deprecated
              firstName: String
            }
          `
          )
        },
        {
          title: "Incorrect",
          code: (
            /* GraphQL */
            `
            type User {
              firstname: String @deprecated(reason: "Use 'firstName' instead")
              firstName: String
            }
          `
          )
        },
        {
          title: "Correct",
          code: (
            /* GraphQL */
            `
            type User {
              firstname: String
                @deprecated(reason: "Use 'firstName' instead", deletionDate: "25/12/2022")
              firstName: String
            }
          `
          )
        }
      ]
    },
    messages: {
      [MESSAGE_REQUIRE_DATE]: 'Directive "@deprecated" must have a deletion date for {{ nodeName }}',
      [MESSAGE_INVALID_FORMAT]: 'Deletion date must be in format "DD/MM/YYYY" for {{ nodeName }}',
      [MESSAGE_INVALID_DATE]: 'Invalid "{{ deletionDate }}" deletion date for {{ nodeName }}',
      [MESSAGE_CAN_BE_REMOVED]: "{{ nodeName }} \u0441an be removed"
    },
    schema
  },
  create(context) {
    return {
      "Directive[name.value=deprecated]"(node) {
        const argName = _optionalChain([context, 'access', _ => _.options, 'access', _2 => _2[0], 'optionalAccess', _3 => _3.argumentName]) || "deletionDate";
        const deletionDateNode = _optionalChain([node, 'access', _4 => _4.arguments, 'optionalAccess', _5 => _5.find, 'call', _6 => _6((arg) => arg.name.value === argName)]);
        if (!deletionDateNode) {
          context.report({
            node: node.name,
            messageId: MESSAGE_REQUIRE_DATE,
            data: {
              nodeName: _utilsjs.getNodeName.call(void 0, node.parent)
            }
          });
          return;
        }
        const deletionDate = _indexjs.valueFromNode.call(void 0, deletionDateNode.value);
        const isValidDate = DATE_REGEX.test(deletionDate);
        if (!isValidDate) {
          context.report({
            node: deletionDateNode.value,
            messageId: MESSAGE_INVALID_FORMAT,
            data: { nodeName: _utilsjs.getNodeName.call(void 0, node.parent) }
          });
          return;
        }
        let [day, month, year] = deletionDate.split("/");
        day = day.padStart(2, "0");
        month = month.padStart(2, "0");
        const deletionDateInMS = Date.parse(`${year}-${month}-${day}`);
        if (Number.isNaN(deletionDateInMS)) {
          context.report({
            node: deletionDateNode.value,
            messageId: MESSAGE_INVALID_DATE,
            data: {
              deletionDate,
              nodeName: _utilsjs.getNodeName.call(void 0, node.parent)
            }
          });
          return;
        }
        const canRemove = Date.now() > deletionDateInMS;
        if (canRemove) {
          const { parent } = node;
          const nodeName = parent.name.value;
          context.report({
            node: parent.name,
            messageId: MESSAGE_CAN_BE_REMOVED,
            data: { nodeName: _utilsjs.getNodeName.call(void 0, parent) },
            suggest: [
              {
                desc: `Remove \`${nodeName}\``,
                fix: (fixer) => fixer.remove(parent)
              }
            ]
          });
        }
      }
    };
  }
};


exports.rule = rule;
