// @graphql-eslint/eslint-plugin@4.4.0/cjs/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _createStarExport(obj) { Object.keys(obj) .filter((key) => key !== "default" && key !== "__esModule") .forEach((key) => { if (exports.hasOwnProperty(key)) { return; } Object.defineProperty(exports, key, {enumerable: true, configurable: true, get: () => obj[key]}); }); }var _indexjs = require('./configs/index.js');
var _parserjs = require('./parser.js');
var _processorjs = require('./processor.js');
var _indexjs3 = require('./rules/index.js');
var _typesjs = require('./types.js'); _createStarExport(_typesjs);
var _utilsjs = require('./utils.js');
const processors = { graphql: _processorjs.processor };
var index_default = {
  parser: _parserjs.parser,
  processor: _processorjs.processor,
  rules: _indexjs3.rules,
  configs: _indexjs.configs
};









exports.configs = _indexjs.configs; exports.default = index_default; exports.parseForESLint = _parserjs.parseForESLint; exports.parser = _parserjs.parser; exports.processors = processors; exports.requireGraphQLOperations = _utilsjs.requireGraphQLOperations; exports.requireGraphQLSchema = _utilsjs.requireGraphQLSchema; exports.rules = _indexjs3.rules;
