// @graphql-eslint/eslint-plugin@4.4.0/cjs/rules/no-root-type/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});var _utilsjs = require('../../utils.js');
const schema = {
  type: "array",
  minItems: 1,
  maxItems: 1,
  items: {
    type: "object",
    additionalProperties: false,
    required: ["disallow"],
    properties: {
      disallow: {
        ..._utilsjs.ARRAY_DEFAULT_OPTIONS,
        items: {
          enum: ["mutation", "subscription"]
        }
      }
    }
  }
};
const rule = {
  meta: {
    type: "suggestion",
    hasSuggestions: true,
    docs: {
      category: "Schema",
      description: "Disallow using root types `mutation` and/or `subscription`.",
      url: "https://the-guild.dev/graphql/eslint/rules/no-root-type",
      requiresSchema: true,
      examples: [
        {
          title: "Incorrect",
          usage: [{ disallow: ["mutation", "subscription"] }],
          code: (
            /* GraphQL */
            `
            type Mutation {
              createUser(input: CreateUserInput!): User!
            }
          `
          )
        },
        {
          title: "Correct",
          usage: [{ disallow: ["mutation", "subscription"] }],
          code: (
            /* GraphQL */
            `
            type Query {
              users: [User!]!
            }
          `
          )
        }
      ],
      configOptions: [{ disallow: ["mutation", "subscription"] }]
    },
    schema
  },
  create(context) {
    const schema2 = _utilsjs.requireGraphQLSchema.call(void 0, "no-root-type", context);
    const disallow = new Set(context.options[0].disallow);
    const rootTypeNames = [
      disallow.has("mutation") && schema2.getMutationType(),
      disallow.has("subscription") && schema2.getSubscriptionType()
    ].filter((v) => !!v).map((type) => type.name).join("|");
    if (!rootTypeNames) {
      return {};
    }
    const selector = `:matches(ObjectTypeDefinition, ObjectTypeExtension) > .name[value=/^(${rootTypeNames})$/]`;
    return {
      [selector](node) {
        const typeName = node.value;
        context.report({
          node,
          message: `Root type \`${typeName}\` is forbidden.`,
          suggest: [
            {
              desc: `Remove \`${typeName}\` type`,
              fix: (fixer) => fixer.remove(node.parent)
            }
          ]
        });
      }
    };
  }
};


exports.rule = rule;
