// parallel-web@0.2.4/resources/beta/beta.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Beta = void 0;
const tslib_1 = require("../../internal/tslib.js");
const resource_1 = require("../../core/resource.js");
const FindallAPI = tslib_1.__importStar(require("./findall.js"));
const findall_1 = require("./findall.js");
const TaskGroupAPI = tslib_1.__importStar(require("./task-group.js"));
const TaskRunAPI = tslib_1.__importStar(require("./task-run.js"));
const task_run_1 = require("./task-run.js");
const headers_1 = require("../../internal/headers.js");
class Beta extends resource_1.APIResource {
    constructor() {
        super(...arguments);
        this.taskRun = new TaskRunAPI.TaskRun(this._client);
        this.taskGroup = new TaskGroupAPI.TaskGroup(this._client);
        this.findall = new FindallAPI.Findall(this._client);
    }
    /**
     * Extracts relevant content from specific web URLs.
     *
     * To access this endpoint, pass the `parallel-beta` header with the value
     * `search-extract-2025-10-10`.
     */
    extract(params, options) {
        const { betas, ...body } = params;
        return this._client.post('/v1beta/extract', {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([
                { 'parallel-beta': [...(betas ?? []), 'search-extract-2025-10-10'].toString() },
                options?.headers,
            ]),
        });
    }
    /**
     * Searches the web.
     *
     * To access this endpoint, pass the `parallel-beta` header with the value
     * `search-extract-2025-10-10`.
     */
    search(params, options) {
        const { betas, ...body } = params;
        return this._client.post('/v1beta/search', {
            body,
            ...options,
            headers: (0, headers_1.buildHeaders)([
                { 'parallel-beta': [...(betas ?? []), 'search-extract-2025-10-10'].toString() },
                options?.headers,
            ]),
        });
    }
}
exports.Beta = Beta;
Beta.TaskRun = task_run_1.TaskRun;
Beta.Findall = findall_1.Findall;
