// parallel-web@0.2.4/resources/beta/findall.mjs
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
import { APIResource } from "../../core/resource.mjs";
import { buildHeaders } from "../../internal/headers.mjs";
import { path } from "../../internal/utils/path.mjs";
export class Findall extends APIResource {
    /**
     * Starts a FindAll run.
     *
     * This endpoint immediately returns a FindAll run object with status set to
     * 'queued'. You can get the run result snapshot using the GET
     * /v1beta/findall/runs/{findall_id}/result endpoint. You can track the progress of
     * the run by:
     *
     * - Polling the status using the GET /v1beta/findall/runs/{findall_id} endpoint,
     * - Subscribing to real-time updates via the
     *   /v1beta/findall/runs/{findall_id}/events endpoint,
     * - Or specifying a webhook with relevant event types during run creation to
     *   receive notifications.
     */
    create(params, options) {
        const { betas, ...body } = params;
        return this._client.post('/v1beta/findall/runs', {
            body,
            ...options,
            headers: buildHeaders([
                { 'parallel-beta': [...(betas ?? []), 'findall-2025-02-01'].toString() },
                options?.headers,
            ]),
        });
    }
    /**
     * Retrieve a FindAll run.
     */
    retrieve(findallID, params = {}, options) {
        const { betas } = params ?? {};
        return this._client.get(path `/v1beta/findall/runs/${findallID}`, {
            ...options,
            headers: buildHeaders([
                { 'parallel-beta': [...(betas ?? []), 'findall-2025-02-01'].toString() },
                options?.headers,
            ]),
        });
    }
    /**
     * Cancel a FindAll run.
     */
    cancel(findallID, params = {}, options) {
        const { betas } = params ?? {};
        return this._client.post(path `/v1beta/findall/runs/${findallID}/cancel`, {
            ...options,
            headers: buildHeaders([
                { 'parallel-beta': [...(betas ?? []), 'findall-2025-02-01'].toString() },
                options?.headers,
            ]),
        });
    }
    /**
     * Add an enrichment to a FindAll run.
     */
    enrich(findallID, params, options) {
        const { betas, ...body } = params;
        return this._client.post(path `/v1beta/findall/runs/${findallID}/enrich`, {
            body,
            ...options,
            headers: buildHeaders([
                { 'parallel-beta': [...(betas ?? []), 'findall-2025-02-01'].toString() },
                options?.headers,
            ]),
        });
    }
    /**
     * Stream events from a FindAll run.
     *
     * Args: request: The Shapi request findall_id: The FindAll run ID last_event_id:
     * Optional event ID to resume from. timeout: Optional timeout in seconds. If None,
     * keep connection alive as long as the run is going. If set, stop after specified
     * duration.
     */
    events(findallID, params = {}, options) {
        const { betas, ...query } = params ?? {};
        return this._client.get(path `/v1beta/findall/runs/${findallID}/events`, {
            query,
            ...options,
            headers: buildHeaders([
                { 'parallel-beta': [...(betas ?? []), 'findall-2025-02-01'].toString(), Accept: 'text/event-stream' },
                options?.headers,
            ]),
            stream: true,
        });
    }
    /**
     * Extend a FindAll run by adding additional matches to the current match limit.
     */
    extend(findallID, params, options) {
        const { betas, ...body } = params;
        return this._client.post(path `/v1beta/findall/runs/${findallID}/extend`, {
            body,
            ...options,
            headers: buildHeaders([
                { 'parallel-beta': [...(betas ?? []), 'findall-2025-02-01'].toString() },
                options?.headers,
            ]),
        });
    }
    /**
     * Transforms a natural language search objective into a structured FindAll spec.
     *
     * Note: Access to this endpoint requires the parallel-beta header.
     *
     * The generated specification serves as a suggested starting point and can be
     * further customized by the user.
     */
    ingest(params, options) {
        const { betas, ...body } = params;
        return this._client.post('/v1beta/findall/ingest', {
            body,
            ...options,
            headers: buildHeaders([
                { 'parallel-beta': [...(betas ?? []), 'findall-2025-02-01'].toString() },
                options?.headers,
            ]),
        });
    }
    /**
     * Retrieve the FindAll run result at the time of the request.
     */
    result(findallID, params = {}, options) {
        const { betas } = params ?? {};
        return this._client.get(path `/v1beta/findall/runs/${findallID}/result`, {
            ...options,
            headers: buildHeaders([
                { 'parallel-beta': [...(betas ?? []), 'findall-2025-02-01'].toString() },
                options?.headers,
            ]),
        });
    }
    /**
     * Get FindAll Run Schema
     */
    schema(findallID, params = {}, options) {
        const { betas } = params ?? {};
        return this._client.get(path `/v1beta/findall/runs/${findallID}/schema`, {
            ...options,
            headers: buildHeaders([
                { 'parallel-beta': [...(betas ?? []), 'findall-2025-02-01'].toString() },
                options?.headers,
            ]),
        });
    }
}
